### DISCUSSION TITLE
How do I secure my database/api in production?

### DISCUSSION THREAD

**Discussion by codepilotsf** [pocketbase/pocketbase#3542] [Q&A]
I feel like I'm missing something really obvious -- but if I'm using PocketBase as the data store for a SvelteKit app, and using the JS SDK to read/write data, what's the best practice to lock it down so that only my app (on the same server) has access? It seems like by default, anyone can run queries against mysite:8090/api and do whatever they want. What am I missing?

**Comment by ganigeorgiev** [comment]
Access control for your data is usually managed via the [Collections API rules](https://pocketbase.io/docs/api-rules-and-filters/).
It is OK PocketBase to be accessible from anyone on the web, just make sure that your API rules are configured and you use a strong admin password.

If you want to "hide" PocketBase behind your node server (aka. only the node server to have access to it) and if both node and PocketBase are deployed on the same server, then you you can use a local address and port not exposed to the internet when starting the executable. But this will require some special route handling to make an exception for your IP so that you can access the Admin UI.

Alternatively you can whitelist only the IPs you want to be able to access the PocketBase instance. This can be done on reverse-proxy level, os firewall or even using a PocketBase middleware like:

```js
// pb_hooks/main.pb.js

routerUse((next) => {
    return (c) => {
        const allowedIPs = ["YOUR_NODE_SERVER_IP", "YOUR_OWN_IP", "..."];

        // extract the client IP (without the port)
        // 
        // note: if you are behind a reverse-proxy, eg. Cloudflare,
        // remoteAddr will be the proxy address and you have to extract
        // the client IP from some trusted header
        // (for Cloudflare that is usually "CF-Connecting-IP")
        const remoteAddr = c.request().remoteAddr; // ip:port
        const lastIndex = remoteAddr.lastIndexOf(":");
        const ip = lastIndex > 0 ? remoteAddr.substring(0, lastIndex) : remoteAddr;

        if (!allowedIPs.includes(ip)) {
            throw new ForbiddenError("You are not allowed to access this resource.")
        }

        return next(c)
    }
})
```

_Edit: update the IP extraction as I forgot that remoteAddr returns also the port (you can skip the port trim if you want to be more precise)._

**Comment by codepilotsf** [comment]
Thanks so much for the detailed reply. Just to make sure I fully understand though, it sounds like you're saying that for users like me looking for a data store for SvelteKit (SSR) loaders/actions, maybe PocketBase isn't an ideal fit out of the box. If I'm using SQLite directly, or even something like MongoDB Atlas, these require direct file access or API tokens – so they're secure by default. With PocketBase, I'll need to be very careful to make sure everyone on the team is locking down every route or add some middleware or something. Do I basically have this right? If so, it seems like there may be an opportunity to expand PocketBase's addressable market by adding an "SSR" mode to somehow make the JS SDK accessible directly instead of over HTTP – or any other scheme which makes it so that your data is locked down by default.

**Reply by ganigeorgiev** [reply]
PocketBase collections are "locked" by default and can be accessed only by admins. More details about the API rules you can find in https://pocketbase.io/docs/api-rules-and-filters/.

If you are not going to use the PocketBase web APIs and want direct access to the SQLite database, then Yes, you wouldn't benefit much from using PocketBase (or any other 3rd party web backend).

**Reply by codepilotsf** [reply]
Got it! Thanks again for your helpful response. For what it's worth, I think my team would definitely benefit from PocketBase's other features like built-in auth, sexy ORM query syntax, GUI, etc. But for an SSR-based app like SvelteKit, it just sorta "feels" wrong to be making HTTP calls to the database. Anyway, just food for thought. I think there may be a lot of folks like me who think PocketBase is soooo close to being a perfect "set of batteries" for the ones that SvelteKit doesn't include. Thanks so much for your hard work on this project. It's really a fantastic tool.
